### PR TITLE
Make UseRefs more memory efficient

### PR DESCRIPTION
**Pull request by Keno**
With #44494, this cuts about 22M allocations (out of 59M) from the
compiler benchmark in #44492. Without #44494, it still reduces
the number of allocations, but not as much. This was originally
optimized in 100666be80f37f79ee709137bd7425e98c33afdd, but the
behavior of our compiler has changed to allow inling the Tuple{UseRef, Int}
into the outer struct, forcing a reallocation on every iteration.

### CODE REVIEW

File: base/compiler/ssair/ir.jl [hunk 1]
```diff
@@ -332,70 +332,65 @@ end
...
-        isassigned(stmt.values, x.op) || return UNDEF_TOKEN
-        return stmt.values[x.op]
+        op > length(stmt.values) && return OOB_TOKEN
+        isassigned(stmt.values, op) || return UNDEF_TOKEN
+        return stmt.values[op]
     else
         return OOB_TOKEN
     end
 end
+@inline getindex(x::UseRef) = _useref_getindex(x.urs.stmt, x.op)
```
Comment (vtjnash) [inline base/compiler/ssair/ir.jl:393]: I had hoped our calling convention would have been enough already to make this inlining awkwardness unnecessary. What causes it to be needed?

Comment (Keno) [reply base/compiler/ssair/ir.jl:393]: We want `UseRef` to be SROA'd, so that `UseRefIterator` can be SROA'd. Without it `UseRefIterator` gets allocated.

**Review by vtjnash — COMMENTED** [review]
It does seem slightly odd that that needs to be mutable, since that implies we eventually need to copy the stmt back to the Instruction steam.

**Comment by Keno** [comment]
> It does seem slightly odd that that needs to be mutable, since that implies we eventually need to copy the stmt back to the Instruction steam.

Yes, that's how this API works. At the end you need to put `urs[]` back into the instruction stream (or whatever else you were modifying).

**Comment by Keno** [comment]
Merging this now - the test for zero allocation will be added in #44557.

**Pull request merged**